File: ThreeSemester/Algo/algo-Lab1-Java/src/Planes.java
import java.util.*;


public class Planes {

    static int graph[][];

    static boolean g[][], used[];

    static int n;


    //type = 0 original edges

    //type = 1 reversed edges

    static void dfs(int v, int type) {

        used[v] = true;

        for (int i = 0; i < n; i++)

            if ((type == 1 ? g[i][v] : g[v][i]) && !used[i]) dfs(i, type);

    }


    static boolean AllVisited() {

        for (int i = 0; i < n; i++)

            if (!used[i]) return false;

        return true;

    }


    public static void main(String[] args) {

        Scanner con = new Scanner(System.in);

        n = con.nextInt();

        graph = new int[n][n];

        for (int i = 0; i < n; i++)

            for (int j = 0; j < n; j++)

                graph[i][j] = con.nextInt();


        g = new boolean[n][n];

        used = new boolean[n];

        int L = 0, R = (int) 1e9;

        while (L < R) {

            int Mid = (L + R) / 2;

            for (int i = 0; i < n; i++)

                for (int j = 0; j < n; j++)

                    g[i][j] = (graph[i][j] <= Mid);


            Arrays.fill(used, false);

            dfs(0, 0);


            boolean flag = false;

            if (AllVisited()) {

                Arrays.fill(used, false);

                dfs(0, 1);

                if (!AllVisited()) flag = true;

            } else flag = true;


            if (flag == false) R = Mid;
            else L = Mid + 1;

        }


        System.out.println(L);

        con.close();

    }

} 

File: sem_3/Algo/LabC/Planes.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<vector<int>> edges, g;
int n;

void dfs(int v);
void dfs1(int v);

int main() {
    cin >> n;
    n++;
    edges.resize(n, vector<int>(n, 0));
    g.resize(n, vector<int>(n, 0));
    for (int i = 1; i < n; i++) {
        for (int j = 1; j < n; j++) {
            cin >> edges[i][j];
        }
    }
    int l = 0;
    int r = 1e9;
    while (l < r) {
        int m = (l + r) / 2;
        for (int i = 1; i < n; i++) {
            for (int j = 1; j < n; j++) {
                if (edges[i][j] <= m) {
                    g[i][j] = 1;
                    continue;
                }
                g[i][j] = 0;
            }
        }
        std::fill(edges[0].begin(), edges[0].end(), 0);
        dfs(1);
        int f = 1;
        for (int i = 1; i < n; i++) {
            if (edges[0][i] == 0) {
                f = 0;
                break;
            }
        }
        if (f == 1) {
            f = 0;
            std::fill(edges[0].begin(), edges[0].end(), 0);
            dfs1(1);
            for (int i = 1; i < n; i++) {
                if (edges[0][i] == 0) {
                    f = 1;
                    break;
                }
            }
        } else {
            f = 1;
        }
        if (f == 0) {
            r = m;
        } else {
            l = m + 1;
        }
    }
    cout << l;
    return 0;
}

void dfs(int v) {
    edges[0][v] = 1;
    for (int i = 1; i < n; i++) {
        if (g[v][i] == 1 && edges[0][i] == 0) {
            dfs(i);
        }
    }
}

void dfs1(int v) {
    edges[0][v] = 1;
    for (int i = 1; i < n; i++) {
        if (g[i][v] == 1 && edges[0][i] == 0) {
            dfs1(i);
        }
    }
}
